src/libpmem2/map_posix.c

```diff
@@ -527,22 +527,33 @@ pmem2_map_new(struct pmem2_map **map_ptr, const struct pmem2_config *cfg,
 				cfg->requested_max_granularity);
 		ERR("%s", err);
 		ret = PMEM2_E_GRANULARITY_NOT_SUPPORTED;
-		goto err_gran_reg;
+		goto err_undo_mapping;
 	}
 
+	/* prepare pmem2_map structure */
+	map = (struct pmem2_map *)pmem2_malloc(sizeof(*map), &ret);
+	if (!map)
+		goto err_undo_mapping;
+
 	map->addr = addr;
 	map->reserved_length = reserved_length;
 	map->content_length = content_length;
 	map->effective_granularity = available_min_granularity;
 	pmem2_set_flush_fns(map);
 	pmem2_set_mem_fns(map);
-	map->reserv = cfg->reserv;
+	map->reserv = rsv;
 	map->source = *src;
 	map->source.value.fd = INVALID_FD; /* fd should not be used after map */
 
 	ret = pmem2_register_mapping(map);
 	if (ret) {
-		goto err_gran_reg;
+		goto err_free_map_struct;
+	}
+
+	if (rsv) {
+		ret = vm_reservation_map_register(rsv, map);
```

Comment:
And this should then should internally release the reservation lock.

Modules (units) should be entirely separate in functionality. When designing a module interface consider its usage in isolation, and make it hard to misuse - just like you would for a public library interface.
In this case, exposing a raw mutex is just asking for deadlocks. The API semantics should be designed in a way that solves the concurrency problem internally, without exposing the implementation.

Reply:
I've made register and unregister functions as lock release functions, i've also added a blank release function in case of errors between *acquire* and desired *release* operation.